crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1902 @@
+    // = note: the enum is complex because of non-unit variant `Leaf`
+    Auto {},
+    AsRoot {},
+    Leaf { index: TreeIndex, side: Side },
+}
+
+#[derive(Copy, Clone, Hash, Debug, PartialEq, Eq, Streamable)]
+pub struct NodeMetadata {
+    // OPT: could save 1-2% of tree space by packing (and maybe don't do that)
+    pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "InternalNode", get_all))]
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, Streamable)]
+pub struct InternalNode {
+    parent: Parent,
+    hash: Hash,
+    left: TreeIndex,
+    right: TreeIndex,
+}
+
+impl InternalNode {
+    pub fn sibling_index(&self, index: TreeIndex) -> Result<TreeIndex, Error> {
+        if index == self.right {
+            Ok(self.left)
+        } else if index == self.left {
+            Ok(self.right)
+        } else {
+            Err(Error::IndexIsNotAChild(index))
+        }
+    }
+}
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "LeafNode", get_all))]
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, Streamable)]
+pub struct LeafNode {
+    parent: Parent,
+    hash: Hash,
+    key: KvId,
+    value: KvId,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Node {
+    Internal(InternalNode),
+    Leaf(LeafNode),
+}
+
+impl Node {
+    fn parent(&self) -> Parent {
+        match self {
+            Node::Internal(node) => node.parent,
+            Node::Leaf(node) => node.parent,
+        }
+    }
+
+    fn set_parent(&mut self, parent: Parent) {
+        match self {
+            Node::Internal(node) => node.parent = parent,
+            Node::Leaf(node) => node.parent = parent,
+        }
+    }
+
+    fn hash(&self) -> Hash {
+        match self {
+            Node::Internal(node) => node.hash,
+            Node::Leaf(node) => node.hash,
+        }
+    }
+
+    fn set_hash(&mut self, hash: &Hash) {
+        match self {
+            Node::Internal(ref mut node) => node.hash = *hash,
+            Node::Leaf(ref mut node) => node.hash = *hash,
+        }
+    }
+
+    #[allow(clippy::trivially_copy_pass_by_ref)]
+    pub fn from_bytes(metadata: &NodeMetadata, blob: &DataBytes) -> Result<Self, Error> {
+        Ok(match metadata.node_type {
+            NodeType::Internal => Node::Internal(streamable_from_bytes_ignore_extra_bytes(blob)?),
+            NodeType::Leaf => Node::Leaf(streamable_from_bytes_ignore_extra_bytes(blob)?),
+        })
+    }
+
+    pub fn to_bytes(&self) -> Result<DataBytes, Error> {
+        let mut base = match self {
+            Node::Internal(node) => node.to_bytes(),
+            Node::Leaf(node) => node.to_bytes(),
+        }
+        .map_err(Error::Streaming)?;
+        assert!(base.len() <= DATA_SIZE);
+        base.resize(DATA_SIZE, 0);
+        Ok(base
+            .as_slice()
+            .try_into()
+            .expect("padding was added above, might be too large"))
+    }
+
+    fn expect_leaf(&self, message: &str) -> LeafNode {
```

Comment:
What is this?

Reply:
It's purpose was to make it easy to express that this should indeed be a leaf variant and to panic if not.  Reviewing the uses there are only two in runtime code.  Those two uses would panic if there was a coding bug resulting in corrupted data or cache.  For example, we get an index from our key -> index mapping and then instead of getting a leaf node at that index (key/value node) we get an internal node.  Or an iterator yielding a block but which only yields leaf blocks.